### ISSUE TITLE
generic variable narrowed to `Never` on `if (not) x` case

### ISSUE THREAD

**Issue by pinterior** [microsoft/pyright#6266]
**Describe the bug**

generic variable `v: T` is narrowed to `Never` on 'else' side of `if v:` or 'then' side of `if not v:`.

**Code or Screenshots**

all `reveal_type` should output T

```python
from typing import TypeVar

T = TypeVar("T")

def f(v: T):
    if v:
        reveal_type(v)
    else:
        reveal_type(v)
    if not v:
        reveal_type(v)
    else:
        reveal_type(v)
```
```
$ pyright gt.py 
gt.py
  gt.py:7:21 - information: Type of "v" is "T@f"
  gt.py:9:21 - information: Type of "v" is "Never"
  gt.py:11:21 - information: Type of "v" is "Never"
  gt.py:13:21 - information: Type of "v" is "T@f"
  gt.py:5:10 - warning: TypeVar "T" appears only once in generic function signature (reportInvalidTypeVarUse)
0 errors, 1 warning, 4 informations 
```

**VS Code extension or command-line**
command-line tool, 1.1.333

**Comment by erictraut** [comment]
Thanks for the bug report. This is a long-standing issue, and it affects not only type variables but instances of any class that derive from `object`, are not `@final` and do not override the `__bool__` method.

```python
def f(v: object):
    if not v:
        reveal_type(v) # "Never", but should be "object"
```

I've modified the "truthy/falsy" type guard logic to handle `object` specially — and not to assume that a value annotated as `object` will always evaluate to truthy. The same applies to type variables that have `object` as an upper bound, as in your example above.

This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.334, which I just published. It will also be included in a future release of pylance.
